File: medbrains/apps/android/app-patient/src/main/kotlin/com/medbrains/patient/PortalApi.kt
package com.medbrains.patient

import androidx.compose.runtime.compositionLocalOf
import com.medbrains.kit.ApiClient
import kotlinx.serialization.Serializable

// The patient portal — the only endpoints this app calls. Every read is
// subject-locked at the backend: the patient comes from the token.

@Serializable data class PortalSession(val token: String, val patient_id: String, val tenant_id: String, val expires_in_hours: Int = 0)
@Serializable data class PortalAppointment(val id: String, val appointment_date: String, val status: String, val department_name: String? = null)
@Serializable data class PortalLabReport(val order_id: String, val test_name: String, val parameter_name: String, val value: String, val unit: String? = null, val normal_range: String? = null, val flag: String? = null, val reported_at: String)
@Serializable data class PortalPrescriptionItem(val prescription_id: String, val drug_name: String, val dosage: String, val frequency: String, val duration: String, val prescribed_at: String)
@Serializable data class PortalInvoice(val id: String, val invoice_number: String, val status: String, val total_amount: String, val paid_amount: String = "0", val balance_due: String, val created_at: String)
@Serializable data class PortalEntitlements(val companion: Boolean = false)
@Serializable private data class OtpRequest(val tenant_code: String, val phone: String)
@Serializable private data class OtpVerify(val tenant_code: String, val phone: String, val code: String)
@Serializable data class Ack(val ok: Boolean = true)

class PortalApi(private val client: ApiClient) {
    /** The reply is the same whether or not the number is registered — do not branch on it. */
    suspend fun requestCode(tenantCode: String, phone: String) { client.call("POST", "/api/portal/auth/request-otp", client.json.encodeToString(OtpRequest.serializer(), OtpRequest(tenantCode, phone))) }
    suspend fun verifyCode(tenantCode: String, phone: String, code: String): PortalSession = client.post("/api/portal/auth/verify", OtpVerify(tenantCode, phone, code))
    suspend fun appointments(): List<PortalAppointment> = client.get("/api/portal/appointments")
    suspend fun labReports(): List<PortalLabReport> = client.get("/api/portal/lab-reports")
    suspend fun prescriptions(): List<PortalPrescriptionItem> = client.get("/api/portal/prescriptions")
    suspend fun bills(): List<PortalInvoice> = client.get("/api/portal/bills")
    suspend fun entitlements(): PortalEntitlements = client.get("/api/portal/entitlements")
}

val LocalPortalApi = compositionLocalOf<PortalApi> { error("PortalApi not provided") }
val LocalHospitalCode = compositionLocalOf { "DEFAULT" }

File: medbrains/apps/ios/MedBrainsPatient/Features/PortalApi.swift
import Foundation
import MedBrainsKit

// The patient portal — the only endpoints this app calls. Every read is
// subject-locked at the backend: the patient comes from the token, so there is
// no id to pass and none to get wrong. Wire shapes mirror
// crates/medbrains-server/src/routes/portal.rs.

struct PortalSession: Decodable, Sendable {
    let token: String
    let patientId: String
    let tenantId: String
    let expiresInHours: Int
}

struct PortalAppointment: Decodable, Identifiable, Sendable {
    let id: String
    let appointmentDate: String
    let status: String
    let departmentName: String?
}

struct PortalLabReport: Decodable, Sendable {
    let orderId: String
    let testName: String
    let parameterName: String
    let value: String
    let unit: String?
    let normalRange: String?
    let flag: String?
    let reportedAt: String
}

struct PortalPrescriptionItem: Decodable, Sendable {
    let prescriptionId: String
    let drugName: String
    let dosage: String
    let frequency: String
    let duration: String
    let prescribedAt: String
}

struct PortalInvoice: Decodable, Identifiable, Sendable {
    let id: String
    let invoiceNumber: String
    let status: String
    let totalAmount: String
    let paidAmount: String
    let balanceDue: String
    let createdAt: String
}

/// What the tenant has licensed for this patient, beyond their record.
struct PortalEntitlements: Decodable, Sendable {
    let companion: Bool
}

private struct OtpRequest: Encodable { let tenantCode: String; let phone: String }
private struct OtpVerify: Encodable { let tenantCode: String; let phone: String; let code: String }

struct PortalApi: Sendable {
    let client: ApiClient

    /// Asks for a sign-in code. The reply is the same whether or not the number
    /// is registered, so this cannot be used to find out who is a patient here.
    func requestCode(tenantCode: String, phone: String) async throws {
        let _: Empty = try await client.request(.post, "/api/portal/auth/request-otp", body: OtpRequest(tenantCode: tenantCode, phone: phone))
    }

    func verifyCode(tenantCode: String, phone: String, code: String) async throws -> PortalSession {
        try await client.request(.post, "/api/portal/auth/verify", body: OtpVerify(tenantCode: tenantCode, phone: phone, code: code))
    }

    func appointments() async throws -> [PortalAppointment] { try await client.request(.get, "/api/portal/appointments") }
    /// Only verified results, and nothing carrying an unacknowledged critical alert — the backend applies both rules.
    func labReports() async throws -> [PortalLabReport] { try await client.request(.get, "/api/portal/lab-reports") }
    func prescriptions() async throws -> [PortalPrescriptionItem] { try await client.request(.get, "/api/portal/prescriptions") }
    func bills() async throws -> [PortalInvoice] { try await client.request(.get, "/api/portal/bills") }
    func entitlements() async throws -> PortalEntitlements { try await client.request(.get, "/api/portal/entitlements") }
}
